test/llvmcall.jl

```diff
@@ -205,6 +205,34 @@ module CcallableRetTypeTest
     @test do_the_call() === 42.0
 end
 
+# Issue #48093 - test that non-external globals are not deduplicated
+@inline shmem() = Base.llvmcall(("""
+        @shmem = internal global [1 x i8] zeroinitializer, align 32
+
+        define i8* @entry() #0 {
+            ret i8* getelementptr inbounds ([1 x i8], [1 x i8]* @shmem, i64 0, i64 0)
+        }
```

Comment:
This test is not really valid though. We are perfectly free to emit any number of copies of llvmcall, provided only that we ensure the user cannot tell how many times it has been emitted because all of them get corrected merged together, as if it was only written exactly once, as it appears in the original source code.

Reply:
The test is checking that after codegen has completed we have emitted two global variables that aren't merged into the same one, not that we have emitted multiple copies that do get merged.

Reply:
Yes, but that specific behavior is wrong. There is only one llvmcall, so it should not be emitting duplicated copies of anything.

Reply:
Can you suggest how to implement this functionality then? We need a way to allocate multiple, distinct, thread-global buffers.

Reply:
Also, since this does fix the issue at hand, I suggest we merge this PR (since switching to LLVM's IR linker seems like a good thing, regardless) and work on a better approach for the GPU back-ends for a future version of Julia? We can always add a note to the test, but until we have a proper solution it seems good to make sure the current workaround doesn't regress again.

Reply:
The test should create multiple llvmcall to demonstrate that they are indeed kept separate if they start separate. E.g. make this:
```julia
macro shmem = :(Base.llvmcall(("""
```